src/lib/rc/ghst.cpp

```diff
@@ -0,0 +1,375 @@
+
+			GHST_VERBOSE("Discarding buffer");
+		}
+
+		return false;
+	}
+
+
+	if (current_frame_position < 3) {
+		// wait until we have the address, length and type
+		return false;
+	}
+
+	// now we have at least the header and the type
+
+	const unsigned current_frame_length = ghst_frame.header.length + sizeof(ghst_frame_header_t);
+
+	if (current_frame_length > sizeof(ghst_frame_t) || current_frame_length < 4) {
+		// frame too long or bogus (frame length should be longer than 4, at least 1 address, 1 length, 1 type, 1 data, 1 crc)
+		// discard everything and go into unsynced state
+		current_frame_position = 0;
+		parser_state = ghst_parser_state_t::unsynced;
+		GHST_DEBUG("Frame too long/bogus (%i, type=%i) -> unsync", current_frame_length, ghst_frame.type);
+		return false;
+	}
+
+	if (current_frame_position < current_frame_length) {
+		// we do not have the full frame yet -> wait for more data
+		GHST_VERBOSE("waiting for more data (%i < %i)", current_frame_position, current_frame_length);
+		return false;
+	}
+
+	bool ret = false;
+
+	// now we have the full frame
+
+	if ((ghst_frame.type >= (uint8_t)ghstFrameType::frameTypeFirst) &&
+	    (ghst_frame.type <= (uint8_t)ghstFrameType::frameTypeLast) &&
+	    (ghst_frame.header.length == GHST_TYPE_DATA_CRC_SIZE)) {
+		const uint8_t crc = ghst_frame.payload[ghst_frame.header.length - 2];
+
+		if (crc == ghst_frame_CRC(ghst_frame)) {
+			const ghstPayloadData_t *const rcChannels = (ghstPayloadData_t *)&ghst_frame.payload;
+			*num_values = MIN(max_channels, 16);
+
+			// all frames contain data from chan1to4
+			if (max_channels > 0) { values[0] = convert_channel_value(rcChannels->chan1to4.chan1 >> 1); }
+
+			if (max_channels > 1) { values[1] = convert_channel_value(rcChannels->chan1to4.chan2 >> 1); }
+
+			if (max_channels > 2) { values[2] = convert_channel_value(rcChannels->chan1to4.chan3 >> 1); }
+
+			if (max_channels > 3) { values[3] = convert_channel_value(rcChannels->chan1to4.chan4 >> 1); }
+
+			if (ghst_frame.type == (uint8_t)ghstFrameType::frameType5to8) {
+				if (max_channels > 4) { values[4] = convert_channel_value(rcChannels->chanA << 3); }
+
+				if (max_channels > 5) { values[5] = convert_channel_value(rcChannels->chanB << 3); }
+
+				if (max_channels > 6) { values[6] = convert_channel_value(rcChannels->chanC << 3); }
+
+				if (max_channels > 7) { values[7] = convert_channel_value(rcChannels->chanD << 3); }
+
+			} else if (ghst_frame.type == (uint8_t)ghstFrameType::frameType9to12) {
+				if (max_channels > 8) { values[8] = convert_channel_value(rcChannels->chanA << 3); }
+
+				if (max_channels > 9) { values[9] = convert_channel_value(rcChannels->chanB << 3); }
+
+				if (max_channels > 10) { values[10] = convert_channel_value(rcChannels->chanC << 3); }
+
+				if (max_channels > 11) { values[11] = convert_channel_value(rcChannels->chanD << 3); }
+
+			} else if (ghst_frame.type == (uint8_t)ghstFrameType::frameType13to16) {
```

Comment:
As channels are reset in https://github.com/PX4/PX4-Autopilot/blob/master/src/drivers/rc_input/RCInput.cpp#L192 and you don't receive all channels in each frames, don't you see these channels toggling values?

Reply:
@bkueng I added tests for ghost. Only channels that are received in a frame are handled and decoded. Channels that are not received in the frame are not handled, so they are presented as 0xffff (because of channel reset that you have mentioned in the previous comment).

Reply:
That is a problem then, see https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/rc_update/rc_update.cpp#L378, it will lead to toggling channels. We need to make sure to keep the previous values for channels not updated (or make sure 0xffff is handled later on).

Reply:
@bkueng I made a solution that you suggested to fix the issue. The previous values for channels that are not updated are now kept. Also, new test data is added.